pkg/scheduler/plugins/deviceshare/device_cache.go

```diff
@@ -0,0 +1,290 @@
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+type nodeDeviceCache struct {
+	lock sync.Mutex
+	// nodeDeviceInfos stores nodeDevice for each node
+	// and uses node name as map key.
+	nodeDeviceInfos map[string]*nodeDevice
+}
+
+type nodeDevice struct {
+	lock        sync.Mutex
+	deviceTotal map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceFree  map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceUsed  map[schedulingv1alpha1.DeviceType]deviceResources
+	allocateSet map[schedulingv1alpha1.DeviceType]*corev1.PodList
+}
+
+// deviceResources is used to present resources per device.
+// we use the minor of device as key
+// "0": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+// "1": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+type deviceResources map[int]corev1.ResourceList
+
+func newNodeDeviceCache() *nodeDeviceCache {
+	return &nodeDeviceCache{
+		nodeDeviceInfos: make(map[string]*nodeDevice),
+	}
+}
+
+func newNodeDevice() *nodeDevice {
+	return &nodeDevice{
+		deviceTotal: make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		deviceFree:  make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		deviceUsed:  make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		allocateSet: make(map[schedulingv1alpha1.DeviceType]*corev1.PodList),
+	}
+}
+
+func (in deviceResources) DeepCopy() deviceResources {
+	if in == nil {
+		return nil
+	}
+	out := deviceResources{}
+	for k, v := range in {
+		out[k] = v
+	}
+	return out
+}
+
+func (n *nodeDeviceCache) getNodeDevice(nodeName string) *nodeDevice {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	return n.nodeDeviceInfos[nodeName]
+}
+
+func (n *nodeDevice) resetDeviceFree(deviceType schedulingv1alpha1.DeviceType) {
+	n.deviceFree[deviceType] = n.deviceTotal[deviceType].DeepCopy()
+	for minor, usedResource := range n.deviceUsed[deviceType] {
+		n.deviceFree[deviceType][minor] = quotav1.SubtractWithNonNegativeResult(
+			n.deviceTotal[deviceType][minor],
+			usedResource)
+	}
+}
+
+func (n *nodeDevice) updateDeviceUsed(deviceType schedulingv1alpha1.DeviceType, allocation *apiext.DeviceAllocation, add bool) {
+	if add {
+		n.deviceUsed[deviceType][int(allocation.Minor)] = quotav1.Add(
+			n.deviceUsed[deviceType][int(allocation.Minor)],
+			allocation.Resources)
+	} else {
+		n.deviceUsed[deviceType][int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+			n.deviceUsed[deviceType][int(allocation.Minor)],
+			allocation.Resources)
+	}
+}
+
+func (n *nodeDevice) tryAllocateDevice(podRequest corev1.ResourceList) (apiext.DeviceAllocations, []error) {
+	var allocateResult apiext.DeviceAllocations
+	var errors []error
+
+	for deviceType := range deviceResourceNames {
```

Comment:
the file looks much better than before. I notice for RDMA and FPGA, only schedulingv1alpha1.FPGA/apiext.KoordFPGA, 
schedulingv1alpha1.RDMA/apiext.KoordRDMA is different, so may be you can package a function of "tryAllocateCommonDevice",  and pass the different device varible to it.  same to isMultipleRDMAPod->isMultipleCommonDevicePod.